src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -368,39 +359,23 @@ protected void engineInit(
           "Unsupported TLen value; must be one of {128, 120, 112, 104, 96}");
     }
 
-    if (this.iv != null
-        && this.key != null
-        && (jceOpMode == Cipher.ENCRYPT_MODE || jceOpMode == Cipher.WRAP_MODE)) {
-      if (Arrays.equals(this.iv, iv)
-          && (encodedKey == null || MessageDigest.isEqual(this.key, encodedKey))) {
-        throw new InvalidAlgorithmParameterException(
-            "Cannot reuse same iv and key for GCM encryption");
-      }
-    }
-
     if (iv == null || iv.length == 0) {
       throw new InvalidAlgorithmParameterException("IV must be at least one byte long");
     }
 
-    switch (jceOpMode) {
-      case Cipher.ENCRYPT_MODE:
-      case Cipher.WRAP_MODE:
-        this.opMode = NATIVE_MODE_ENCRYPT;
-        break;
-      case Cipher.DECRYPT_MODE:
-      case Cipher.UNWRAP_MODE:
-        this.opMode = NATIVE_MODE_DECRYPT;
-        break;
-      default:
-        throw new InvalidAlgorithmParameterException("Unsupported cipher mode " + jceOpMode);
+    this.sameKey = this.key != null && MessageDigest.isEqual(this.key, encodedKey);
```

Comment:
You don't get anything from comparing the MessageDigest, just compare the keys directly. Computing the hashes still requires traversing both keys and doing math on top of that. We also risk the odd issue of collision by comparing like this.

Reply:
I think the naming is confusing here. [MessageDigest.isEquals](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/security/MessageDigest.java#L471) is same as Arrays.equals, except it's constant time for both equal and not equal cases.